File: app/src/androidTest/java/org/jdc/template/CustomMatchers.java
package org.jdc.template;

import android.view.View;

import org.hamcrest.Description;

import androidx.recyclerview.widget.RecyclerView;
import androidx.test.espresso.matcher.BoundedMatcher;

public class CustomMatchers {
    public static BoundedMatcher<View, RecyclerView> recyclerViewWithItemCount(final int expectedCount) {
        return new BoundedMatcher<View, RecyclerView>(RecyclerView.class) {
            private int actualCount;

            @Override
            protected boolean matchesSafely(RecyclerView view) {
                actualCount = view.getAdapter().getItemCount();
                return  actualCount == expectedCount;
            }

            @Override
            public void describeTo(final Description description) {
                description.appendText("RecyclerView Count expected: [" + expectedCount + "] actual: [" + actualCount + "]");
            }
        };
    }

}

File: app/src/androidTest/java/org/jdc/template/CustomMatchers.kt
package org.jdc.template

import android.view.View
import androidx.recyclerview.widget.RecyclerView
import androidx.test.espresso.matcher.BoundedMatcher
import org.hamcrest.Description

object CustomMatchers {
    fun recyclerViewWithItemCount(expectedCount: Int): BoundedMatcher<View?, RecyclerView> {
        return object : BoundedMatcher<View?, RecyclerView>(RecyclerView::class.java) {
            private var actualCount = 0
            override fun matchesSafely(view: RecyclerView): Boolean {
                actualCount = view.adapter?.itemCount ?: -1
                return actualCount == expectedCount
            }

            override fun describeTo(description: Description) {
                description.appendText("RecyclerView Count expected: [$expectedCount] actual: [$actualCount]")
            }
        }
    }
}